src/zen/websocket_server.cpp

```diff
@@ -0,0 +1,1360 @@
+
+double GetNetworkDifficulty(const CBlockIndex * pindex);
+
+static boost::shared_ptr<WsNotificationInterface> wsNotificationInterface;
+static std::list< boost::shared_ptr<WsHandler> > listWsHandler;
+
+std::atomic<bool> exit_ws_thread{false};
+boost::thread ws_thread;
+std::mutex wsmtx;
+
+static void dumpUniValueError(const UniValue& error, std::string& outMsg)
+{
+    UniValue errCode = find_value(error, "code");
+    UniValue errMsg  = find_value(error, "message");
+    std::string strPrint = errCode.isNull() ? "" : "error code: " + errCode.getValStr() + "; ";
+    if (errMsg.isStr())
+    {
+        outMsg = errMsg.get_str();
+    }
+    LogPrint("ws", "%s():%d - JSON error: %s\n", __func__, __LINE__, (strPrint + outMsg));
+}
+
+static std::string findFieldValue(const std::string& field, const UniValue& request)
+{
+    const UniValue& clientvalue = find_value(request, field);
+    if (clientvalue.isStr()) {
+        return clientvalue.get_str();
+    } else if (clientvalue.isNum()) {
+        return std::to_string(clientvalue.get_int());
+    } else if (clientvalue.isNull()) {
+        return "";
+    }
+    return "";
+}
+
+class WsNotificationInterface: public CValidationInterface
+{
+protected:
+    virtual void UpdatedBlockTip(const CBlockIndex *pindex) {
+        ws_updatetip(pindex);
+    };
+    virtual void MempoolChanged() {
+        ws_updatemempool();
+    };
+    virtual void PeersChanged() {
+        ws_updatepeer();
+    };
+public:
+    ~WsNotificationInterface()
+    {
+        LogPrint("ws", "%s():%d - called this=%p\n", __func__, __LINE__, this);
+    }
+};
+
+class WsEvent
+{
+public:
+    enum WsEventType {
+        UPDATE_TIP = 0,
+        UPDATE_MEMPOOL = 1,
+		UPDATE_PEER = 2,
+        EVT_UNDEFINED = 0xff
+    };
+    enum WsRequestType {
+        GET_SINGLE_BLOCK = 0,
+        GET_MULTIPLE_BLOCK_HASHES = 1,
+        GET_NEW_BLOCK_HASHES = 2,
+		GET_MEMPOOL_TXS = 4,
+        GET_RAW_MEMPOOL = 5,
+		GET_NODE_STATIC_CONFIG = 6,
+		GET_ESTIMATE_FEE = 7,
+        REQ_UNDEFINED = 0xff
+    };
+
+    enum WsMsgType {
+        MSG_EVENT = 0,
+        MSG_REQUEST = 1,
+        MSG_RESPONSE = 2,
+        MSG_ERROR = 3,
+        MSG_UNDEFINED = 0xff
+    };
+
+    explicit WsEvent(WsMsgType xn): type(xn), payload(UniValue::VOBJ)
+    {
+        payload.push_back(Pair("msgType", type));
+    }
+    WsEvent & operator=(const WsEvent& ws) = delete;
+    WsEvent(const WsEvent& ws) = delete;
+
+    UniValue* getPayload() {
+        return &payload;
+    }
+
+private:
+    WsMsgType type;
+    UniValue payload;
+};
+
+
+
+class WsHandler
+{
+private:
+    boost::shared_ptr< websocket::stream<tcp::socket>> localWs;
+    boost::lockfree::queue<WsEvent*, boost::lockfree::capacity<1024>> wsq;
+    std::atomic<bool> exit_rwhandler_thread_flag { false };
+
+    void sendBlockEvent(int height, const std::string& strHash, const std::string& blockHex, WsEvent::WsEventType eventType, const CBlockIndex * pindex)
```

Comment:
Consider using a const std::string& chainWorkStr instead of passing the pindex.
This would imply also changing the interface of send_tip_update().
The reason is that we already are passing other parameters such as: pindex->nHeight, pindex->GetBlockHash()
therefore we might as well pass pindex->nChainWork.GetHex().